src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java

```diff
@@ -163,47 +164,94 @@ public static long mismatch(AbstractMemorySegmentImpl src, long srcFromOffset, l
         src.checkAccess(srcFromOffset, srcBytes, true);
         dst.checkAccess(dstFromOffset, dstBytes, true);
 
-        final long bytes = Math.min(srcBytes, dstBytes);
+        final long length = Math.min(srcBytes, dstBytes);
         final boolean srcAndDstBytesDiffer = srcBytes != dstBytes;
 
-        if (bytes == 0) {
+        if (length == 0) {
             return srcAndDstBytesDiffer ? 0 : -1;
-        } else if (bytes < NATIVE_THRESHOLD_MISMATCH) {
-            return mismatch(src, srcFromOffset, dst, dstFromOffset, 0, (int) bytes, srcAndDstBytesDiffer);
+        } else if (length < NATIVE_THRESHOLD_MISMATCH) {
+            return mismatch(src, srcFromOffset, dst, dstFromOffset, 0, (int) length, srcAndDstBytesDiffer);
         } else {
             long i;
             if (SCOPED_MEMORY_ACCESS.getByte(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset) !=
                     SCOPED_MEMORY_ACCESS.getByte(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset)) {
                 return 0;
             }
-            i = AbstractMemorySegmentImpl.vectorizedMismatchLargeForBytes(src.sessionImpl(), dst.sessionImpl(),
+            i = vectorizedMismatchLargeForBytes(src.sessionImpl(), dst.sessionImpl(),
                     src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset,
                     dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset,
-                    bytes);
+                    length);
             if (i >= 0) {
                 return i;
             }
             final long remaining = ~i;
             assert remaining < 8 : "remaining greater than 7: " + remaining;
-            i = bytes - remaining;
+            i = length - remaining;
             return mismatch(src, srcFromOffset + i, dst, dstFromOffset + i, i, (int) remaining, srcAndDstBytesDiffer);
         }
     }
 
+    // Mismatch is handled in chunks of 64 (unroll of eight 8s), 8, 4, 2, and 1 byte(s).
     @ForceInline
     private static long mismatch(AbstractMemorySegmentImpl src, long srcFromOffset,
                                  AbstractMemorySegmentImpl dst, long dstFromOffset,
-                                 long start, int bytes, boolean srcAndDstBytesDiffer) {
-        final int limit = (int) (bytes & (NATIVE_THRESHOLD_MISMATCH - 8));
+                                 long start, int length, boolean srcAndDstBytesDiffer) {
         int offset = 0;
-        for (; offset < limit; offset += 8) {
+        // Currently, we do not benefit from super-word optimization on Aarch64 so
+        // instead we manually unroll the loop.
+        // This gives about 20% performance increase for large values of `length`.
+        // On non-Aarch64 architectures, the unroll code will be eliminated at compile time.
+        if (Architecture.isAARCH64() && NATIVE_THRESHOLD_MISMATCH > 64) {
```

Comment:
I'm a bit opposed to this - as it goes in the direction to add a lot of transient complexity when in reality the underlying issue is that aarch64 mismatch intrinsics should be fixed. Tinkering with thresholds is borderline, but still acceptable - having different implementations one per platform starts to look "more wrong".

Reply:
In other words, I don't think the goal of this (and related) PR is "improve mismatch so that it blows other alternatives - like Unsafe, or array" out of the water - as much as it is "make sure that using MemorySegment::mismatch is competitive with other offerings".

Reply:
Then, an interesting part of these PRs is that we have uncovered quite a lot of issues both with our intrinsics (e.g. `fill` is not vectorized and works badly on Windows, mismatch works poorly on aarch64) *and* missed optimization opportunities - e.g. "short" segment loops are not optimized as tightly as they could. But it is not the job of these PRs to fix all these issues. The main focus remain: for small sizes it is not worth going down intrinsics-lane. Let's stick to it (in the interest of keeping the review focused).

Reply:
Yepp. So, let us keep these tricks up our sleeves and then maybe come back with a new PR.